operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java

```diff
@@ -0,0 +1,137 @@
+   * in the proposed domain resource can be honored, otherwise, returns false.
+   *
+   * @param existingDomain domain that needs to be validated against.
+   * @param proposedDomain domain that needs to be validated.
+   * @return true if valid, otherwise false
+   */
+  public static boolean validateDomain(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    LOGGER.fine("Validating domain " + proposedDomain + " against " + existingDomain);
+    return unchanged(existingDomain, proposedDomain)
+        || validReplicas(proposedDomain)
+        || shouldIntrospect(existingDomain, proposedDomain);
+  }
+
+  private static boolean unchanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return existingDomain == proposedDomain || specEquals(existingDomain, proposedDomain);
+  }
+
+  private static boolean specEquals(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return Optional.of(existingDomain)
+        .map(Domain::getSpec)
+        .map(s -> isProposedSpecUnchanged(s, proposedDomain))
+        .orElse(false);
+  }
+
+  private static boolean isProposedSpecUnchanged(@NotNull DomainSpec existingSpec, @NotNull Domain proposedDomain) {
+    return existingSpec == proposedDomain.getSpec() || Objects.equals(existingSpec, proposedDomain.getSpec());
+  }
+
+  private static boolean shouldIntrospect(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return !Objects.equals(existingDomain.getIntrospectVersion(), proposedDomain.getIntrospectVersion())
+        || imagesChanged(existingDomain, proposedDomain);
+  }
+
+  private static boolean imagesChanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    if (!Objects.equals(existingDomain.getDomainHomeSourceType(), proposedDomain.getDomainHomeSourceType())) {
+      return true;
+    }
+    switch (proposedDomain.getDomainHomeSourceType()) {
+      case IMAGE:
+        return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+      case FROM_MODEL:
+        if (noAuxiliaryImagesConfigured(existingDomain) && noAuxiliaryImagesConfigured(proposedDomain)) {
+          return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+        } else {
+          return noAuxiliaryImagesConfigured(existingDomain)
+              || noAuxiliaryImagesConfigured(proposedDomain)
+              || !isEqualCollection(existingDomain.getAuxiliaryImages(), proposedDomain.getAuxiliaryImages());
+        }
+      default:
+        return false;
+    }
+  }
+
+  private static String getImage(@NotNull Domain existingDomain) {
+    return Optional.of(existingDomain).map(Domain::getSpec).map(DomainSpec::getImage).orElse(null);
+  }
+
+  private static boolean noAuxiliaryImagesConfigured(@NotNull Domain domain) {
+    return domain.getAuxiliaryImages() == null;
+  }
+
+  private static boolean validReplicas(@NotNull Domain domain) {
```

Comment:
The name of this method and the next are unclear. It seems to me that it returns true if all clusters in the domain have valid replica counts. If so, a better name for this one might be `allReplicaCountsValid` and for the next `isReplicaCountValid`.